src/Artifacts.jl

```diff
@@ -0,0 +1,594 @@
+disk, and retains its integrity. 
+"""
+function verify_artifact(hash::SHA1)
+    if !artifact_exists(hash)
+        return false
+    end
+
+    return hash.bytes == tree_hash(artifact_path(hash))
+end
+
+"""
+    archive_artifact(hash::SHA1, tarball_path::String)
+
+Archive an artifact into a tarball stored at `tarball_path`, returns the SHA256
+of the resultant tarball.  Throws an error if the artifact does not exist.
+"""
+function archive_artifact(hash::SHA1, tarball_path::String)
+    if !artifact_exists(hash)
+        error("Unable to archive artifact $(bytes2hex(hash.bytes)): does not exist!")
+    end
+
+    # Package it up
+    package(artifact_path(hash), tarball_path)
+
+    # Calculate its sha256 and return that
+    return open(tarball_path, "r") do io
+        return bytes2hex(sha256(io))
+    end
+end
+
+
+"""
+    unpack_platform(entry::Dict, name::String, artifacts_toml::String)
+
+Given an `entry` for the artifact named `name`, located within the file `artifacts_toml`,
+returns the `Platform` object that this entry specifies.  Returns `nothing` on error.
+"""
+function unpack_platform(entry::Dict, name::String, artifacts_toml::String)
+    if !haskey(entry, "os")
+        @warn("Invalid Artifacts.toml at '$(artifacts_toml)': platform-specific artifact entry '$name' missing 'os' key")
+        return nothing
+    end
+
+    if !haskey(entry, "arch")
+        @warn("Invalid Artfiact.toml at '$(artifacts_toml)': platform-specific artifact entrty '$name' missing 'arch' key")
+        return nothing
+    end
+
+    # Helpers to pull out `Symbol`s and `VersionNumber`s while preserving `nothing`.
+    nosym(x::Nothing) = nothing
+    nosym(x) = Symbol(lowercase(x))
+    nover(x::Nothing) = nothing
+    nover(x) = VersionNumber(x)
+
+    # First, extract OS; we need to build a mapping here
+    os_map = Dict(
+        "windows" => Windows,
+        "macos" => MacOS,
+        "freebsd" => FreeBSD,
+        "linux" => Linux,
+    )
+    P = get(os_map, lowercase(entry["os"]), UnknownPlatform)
+
+    # Next, architecture, libc, libgfortran version and cxxabi (if given)
+    arch = nosym(get(entry, "arch", nothing))
+    libc = nosym(get(entry, "libc", nothing))
+    libgfortran_version = nover(get(entry, "libgfortran_version", nothing))
+    libstdcxx_version = nover(get(entry, "libstdcxx_version", nothing))
+    cxxstring_abi = nosym(get(entry, "cxxstring_abi", nothing))
+
+    # Construct the actual Platform object
+    return P(arch;
+        libc=libc,
+        compiler_abi=CompilerABI(
+            libgfortran_version=libgfortran_version,
+            libstdcxx_version=libstdcxx_version,
+            cxxstring_abi=cxxstring_abi
+        ),
+    )
+end
+
+function pack_platform!(meta::Dict, p::Platform)
+    os_map = Dict(
+        Windows => "windows",
+        MacOS => "macos",
+        FreeBSD => "freebsd",
+        Linux => "linux",
+    )
+    meta["os"] = os_map[typeof(p)]
```

Comment:
Maybe break the above into very short functions?
`name(p::Windows)="windows"` etc